### PR TITLE
Fold some casts early in importer

### PR DESCRIPTION
**Pull request by EgorBo**
A small CQ issue reported by the community. Example:
```cs
public static void Test() => Callee();

[MethodImpl(MethodImplOptions.AggressiveInlining)]
public static void Callee()
{
    nint* ptr = stackalloc nint[2];
    Consume(ptr);
}

[MethodImpl(MethodImplOptions.NoInlining)]
private static void Consume(IntPtr* ptr) {}
```
Current codegen:
```asm
; Assembly listing for method Cd:Test() (FullOpts)
       sub      rsp, 40
       call     [Cd:Callee()]
       nop      
       add      rsp, 40
       ret      
; Total bytes of code 16
```
^ JIT refuses to inline `Consume` due to "unknown size stackalloc". But it's "unknown" only because importer didn't fold it
New codegen:
```asm
; Assembly listing for method Cd:Test() (FullOpts)
       sub      rsp, 72
       mov      rax, 0x9ABCDEF012345678
       mov      qword ptr [rsp+0x40], rax
       lea      rcx, [rsp+0x28]
       call     [Cd:Consume(ulong)]
       mov      rcx, 0x9ABCDEF012345678
       cmp      qword ptr [rsp+0x40], rcx
       je       SHORT G_M42256_IG04
       call     CORINFO_HELP_FAIL_FAST
G_M42256_IG04:
       nop      
       add      rsp, 72
       ret      
; Total bytes of code 58
```

### CODE REVIEW

**Comment by EgorBo** [comment]
@dotnet/jit-contrib PTAL, simple change, fixes the snippet ^. [Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=821357&view=ms.vss-build-web.run-extensions-tab)

**Review by AndyAyersMS — APPROVED** [review]

File: src/coreclr/jit/importer.cpp [hunk 1]
```diff
@@ -5569,13 +5572,21 @@ var_types Compiler::impGetByRefResultType(genTreeOps oper, bool fUnsigned, GenTr
...
             // insert an explicit upcast
-            op2 = *pOp2 = gtNewCastNode(TYP_I_IMPL, op2, fUnsigned, TYP_I_IMPL);
+            op2 = gtNewCastNode(TYP_I_IMPL, op2, fUnsigned, TYP_I_IMPL);
+        }
+
+        if (opts.OptimizationEnabled())
+        {
+            op1 = gtFoldExpr(op1);
+            op2 = gtFoldExpr(op2);
         }
```
Comment (tannergooding) [inline src/coreclr/jit/importer.cpp:5587]: Most other places in import don't guard, they just call `gtFoldExpr` and let the `!opts.Tier0OptimizationEnabled()` early exit, do we want to do the same here?

Presumably there are opportunities for dead code elimination and similar that we may want to allow?

-- If not, can we add a comment on why we're differing here?

Comment (tannergooding) [reply src/coreclr/jit/importer.cpp:5587]: Also, are we wanting to fold here even if we didn't introduce a new cast node?

**Review by tannergooding — APPROVED** [review]

File: src/coreclr/jit/importer.cpp (hunk 1 above)
Comment (EgorBo) [reply src/coreclr/jit/importer.cpp:5587]: > Also, are we wanting to fold here even if we didn't introduce a new cast node?

it seems it introduced some improvements so I didn't limit to newly introduced casts

> Most other places in import don't guard, they just call

I didn't initially, but it introduced some size regressions for tier0 so I just disabled it 🙂

**Pull request merged**